### DISCUSSION TITLE
Orphan scheduler process with supervisor

### DISCUSSION THREAD

**Discussion by cypx** [apache/airflow#27985] [Q&A]
Hello,

I use supervisor to manage Airflow services and I get one  `airflow scheduler -- DagFileProcessorManager`  orphan process each time I stop/start or restart scheduler.

I get the same problem with a minimal Docker install.

- supervisord.conf

```
; supervisor config file

[unix_http_server]
file=/var/run/supervisor.sock   ; (the path to the socket file)
chmod=0700                       ; sockef file mode (default 0700)

[supervisord]
logfile=/var/log/supervisord.log ; (main log file;default $CWD/supervisord.log)
pidfile=/var/run/supervisord.pid ; (supervisord pidfile;default supervisord.pid)
childlogdir=/tmp            ; ('AUTO' child log dir, default $TEMP)

; the below section must remain in the config file for RPC
; (supervisorctl/web interface) to work, additional interfaces may be
; added by defining them in separate rpcinterface: sections
[rpcinterface:supervisor]
supervisor.rpcinterface_factory = supervisor.rpcinterface:make_main_rpcinterface

[supervisorctl]
serverurl=unix:///var/run/supervisor.sock ; use a unix:// URL  for a unix socket

; The [include] section can just contain the "files" setting.  This
; setting can list multiple files (separated by whitespace or
; newlines).  It can also contain wildcards.  The filenames are
; interpreted as relative to this file.  Included files *cannot*
; include files themselves.

[program:airflow-scheduler]
command=airflow scheduler --skip-serve-logs
stopsignal=USR2
stopasgroup=true
killasgroup=true
autorestart=true
```

- Dockerfile:

```
FROM python:3.10
RUN apt-get update && apt-get install -y libsasl2-dev
RUN pip3 install apache-airflow==2.3.4
RUN pip3 install apache-airflow[hive]==2.3.4
RUN pip3 install supervisor==4.2.4
RUN airflow db init
COPY supervisord.conf /etc/supervisord.conf
```

Build, run container and launch supervisor

```bash
sh-5.1$ docker build . -t airflow
Sending build context to Docker daemon  4.096kB
Step 1/7 : FROM python:3.10
 ---> 465483cdaa4e
Step 2/7 : RUN apt-get update && apt-get install -y libsasl2-dev
 ---> Using cache
 ---> e0b516124c9d
Step 3/7 : RUN pip3 install apache-airflow==2.3.4
 ---> Using cache
 ---> 96aed668a5de
Step 4/7 : RUN pip3 install apache-airflow[hive]==2.3.4
 ---> Using cache
 ---> 6440a644f574
Step 5/7 : RUN pip3 install supervisor==4.2.4
 ---> Using cache
 ---> 8cd05d1674b6
Step 6/7 : RUN airflow db init
 ---> Using cache
 ---> 07decd0c3a03
Step 7/7 : COPY supervisord.conf /etc/supervisord.conf
 ---> Using cache
 ---> 7e2b8f873fb5
Successfully built 7e2b8f873fb5
Successfully tagged airflow:latest
sh-5.1$ docker run --rm -it airflow /bin/bash
root@8eb724716a0a:/# supervisord -c /etc/supervisord.conf
root@8eb724716a0a:/# supervisorctl status
airflow-scheduler                RUNNING   pid 9, uptime 0:00:05
root@8eb724716a0a:/# supervisorctl stop airflow-scheduler
airflow-scheduler: stopped
root@8eb724716a0a:/# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0   6056  3708 pts/0    Ss   13:21   0:00 /bin/bash
root           8 11.4  0.1  32260 23540 ?        Ss   13:21   0:10 /usr/local/bin/python /usr/local/bin/supervisord
root          21  0.1  1.1 647720 179472 ?       S    13:21   0:00 airflow scheduler -- DagFileProcessorManager
root          58  0.0  0.0   8652  3208 pts/0    R+   13:23   0:00 ps aux
root@8eb724716a0a:/# supervisorctl start airflow-scheduler

```

Now, if I stop airflow-scheduler an `airflow scheduler -- DagFileProcessorManager` persist.  
If I restart airflow-scheduler I get two  `airflow scheduler -- DagFileProcessorManager`  process

```bash
root@8eb724716a0a:/# supervisorctl start airflow-scheduler
airflow-scheduler: started
root@8eb724716a0a:/# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0   6056  3708 pts/0    Ss   13:21   0:00 /bin/bash
root           8  9.9  0.1  32260 23540 ?        Ss   13:21   0:10 /usr/local/bin/python /usr/local/bin/supervisord
root          21  0.1  1.1 647720 179472 ?       S    13:21   0:00 airflow scheduler -- DagFileProcessorManager
root          60 31.3  1.2 648260 201644 ?       S    13:23   0:01 /usr/local/bin/python /usr/local/bin/airflow scheduler --skip-serve-logs
root          72  0.8  1.1 648000 184608 ?       S    13:23   0:00 airflow scheduler -- DagFileProcessorManager
root          78  0.0  0.0   8652  3184 pts/0    R+   13:23   0:00 ps aux
```

I try different value for `stopsignal` (QUIT, KILL...) without any difference.
Is it a misconfiguration or a bug?

**Comment by potiuk** [comment]
It makes very little sense to run supervisord in docker container. You should manage your container start/stop from outside - using any other kind of supervision (docker-compoe, kubernetes, swarm or even manage your containers separately. 

Containers by definition are there to run individual processes - not to manage the whole suite of processes using supervisord. And airflow had never been tested with it - you need to make sure you know what you are doing when you are running Airlfow this way. While some of the docker docs suggest it as a "heavyweight" optionm we do not recommend it. And you are mostly on your own here, we do not have any best practices or expectations for it to work (this is the "heavyweight" part that Docker docs do not explain you.

You need to debug and find out what you shoudl be doing to achieve what you want, but you might look more into some of our docs and maybe you will find the right approach:

* https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html - look at various options connected with daemonizing airflow. Running non-daemon airflow under supervisord might be a problem
* look at signal propagation in Airflow image documentation https://airflow.apache.org/docs/docker-stack/entrypoint.html#signal-propagation - you might somehow hit similar problems as described there, generally speaking - when you are running scheduler you should make sure the signals are sent to process GROUP not to the process (and this is how dumb-init in our image works). I think supervisord supports it - but you shouuld check how. This is differnet with celery because celery manages it's forked process on its own and handles "warm shutdown" this way.

**Comment by cypx** [comment]
Thanks for your response but Docker is not available on the target system.

At the end, I was able to solve the problem by using TERM signal, it's the default one of supervisor but following some examples (and/or the doc) I start with USR2 because it seems recommended.

**Reply by potiuk** [reply]
Ah. Your example shown that you are using docker to run it - so that was my understanding you want to run supervisord in docker (this is what your example shows). But yeah. I would never thought one would want to use SIGUSR2 - this is a custom signal that is application specific, that is rarely used for something like application lifecycle managemebt. Usually (and this is best practice) all the tools are using signal escalation with timeouts (look it up): SIGTERM, SIGINT, SIGSTOP, SIGKILL and usually, when you have a  process that runs a lot of forked processes you send the signal to pricess group. The latter is usually not needed, because - unless the children of the process are deamonising themselves and detaching from the main process, they will be killed

Actually, I know what is the problem you had. It comes from not understanding how docker and init process interact.

When you run `docker run --rm -it airflow /bin/bash` - you set bash as an "init" process (PID=1). And bash is very bad as an init process, because it does not propagate signals and does not reap zombies. Look for example here: https://stackoverflow.com/questions/49162358/docker-init-zombies-why-does-it-matter

In your case what happened - when you killed main scheduler process, the forked processes were detached and got the bash process of yours (PID=1) as a parent. Bash does not know how to reap zombies that are attached to it and as the result they will keep running. This is what you observed.

If you want to do a proper init, in modern versions of docker you could add `--init` flag:

```
docker run --init --rm -it airflow /bin/bash
```

This one starts `tini` as your init process and tini both - propagates signals and knows how to reap zombies.
